File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/HotRodEncoder.scala
package org.infinispan.server.hotrod

import io.netty.buffer.ByteBuf
import io.netty.util.internal.PlatformDependent
import org.infinispan.manager.EmbeddedCacheManager
import org.infinispan.commons.util.Util
import io.netty.handler.codec.MessageToByteEncoder
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelHandler.Sharable
import org.infinispan.commons.logging.LogFactory
import org.infinispan.server.hotrod.Events.Event
import org.infinispan.server.hotrod.OperationStatus._
import org.infinispan.server.hotrod.logging.JavaLog

/**
 * Hot Rod specific encoder.
 *
 * @author Galder Zamarreño
 * @since 4.1
 */
@Sharable
class HotRodEncoder(cacheManager: EmbeddedCacheManager, server: HotRodServer)
        extends MessageToByteEncoder[Any](PlatformDependent.directBufferPreferred()) with Constants {

   private lazy val isClustered: Boolean = cacheManager.getCacheManagerConfiguration.transport.transport != null
   private lazy val addressCache: AddressCache =
      if (isClustered) cacheManager.getCache(server.getConfiguration.topologyCacheName) else null

   val log = LogFactory.getLog(getClass, classOf[JavaLog])
   private val isTrace = log.isTraceEnabled

   def encode(ctx: ChannelHandlerContext, msg: Any, buf: ByteBuf): Unit = {
      try {
         log.tracef("Encode msg %s", msg)

         msg match {
            case r: Response =>
               val encoder = getEncoder(r.version)
               try {
                  r.version match {
                     case ver if Constants.isVersionKnown(ver) =>
                        encoder.writeHeader(r, buf, addressCache, server)
                     // if error before reading version, don't send any topology changes
                     // cos the encoding might vary from one version to the other
                     case 0 => encoder.writeHeader(r, buf, null, server)
                  }

                  encoder.writeResponse(r, buf, cacheManager, server)
               }
               catch {
                  case t: Throwable =>
                     log.errorWritingResponse(r.messageId, t)
                     buf.clear() // reset buffer
                  val error = new ErrorResponse(r.version, r.messageId, r.cacheName, r.clientIntel, ServerError, r.topologyId, t.toString)
                     encoder.writeHeader(error, buf, addressCache, server)
                     encoder.writeResponse(error, buf, cacheManager, server)
               }
            case e: Event =>
               val encoder = getEncoder(e.version)
               encoder.writeEvent(e, buf)
            case None =>
               // Do nothing
            case _ =>
               log.errorUnexpectedMessage(msg)
         }

         if (isTrace)
            log.tracef("Write buffer contents %s to channel %s",
               Array(Util.hexDump(buf.nioBuffer), ctx.channel).map(_.asInstanceOf[AnyRef]) : _*)
      }
      catch {
         case t: Throwable =>
            log.errorEncodingMessage(msg, t)
            throw t
      }
   }

   private def getEncoder(version: Byte): AbstractVersionedEncoder = {
      version match {
         case ver if Constants.isVersion2x(ver) => Encoder2x
         case ver if Constants.isVersion13(ver) => Encoders.Encoder13
         case ver if Constants.isVersion12(ver) => Encoders.Encoder12
         case ver if Constants.isVersion11(ver) => Encoders.Encoder11
         case ver if Constants.isVersion10(ver) => Encoders.Encoder10
         case 0 => Encoder2x
      }
   }

}

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/HotRodEncoder.java
package org.infinispan.server.hotrod;

import org.infinispan.Cache;
import org.infinispan.commons.logging.LogFactory;
import org.infinispan.manager.EmbeddedCacheManager;
import org.infinispan.remoting.transport.Address;
import org.infinispan.server.hotrod.counter.listener.ClientCounterEvent;
import org.infinispan.server.hotrod.logging.Log;

import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandler.Sharable;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.MessageToByteEncoder;
import io.netty.util.internal.PlatformDependent;

/**
 * Hot Rod specific encoder.
 *
 * @author Galder Zamarreño
 * @since 4.1
 */
@Sharable
class HotRodEncoder extends MessageToByteEncoder<Object> {
   private final static Log log = LogFactory.getLog(HotRodEncoder.class, Log.class);
   private final static boolean isTrace = log.isTraceEnabled();
   private final EmbeddedCacheManager cacheManager;
   private final HotRodServer server;
   private final boolean isClustered;
   private Cache<Address, ServerAddress> addressCache;

   HotRodEncoder(EmbeddedCacheManager cacheManager, HotRodServer server) {
      super(PlatformDependent.directBufferPreferred());
      this.cacheManager = cacheManager;
      this.server = server;
      isClustered = cacheManager.getCacheManagerConfiguration().transport().transport() != null;
   }

   private Cache<Address, ServerAddress> getAddressCache() {
      if (addressCache == null) {
         addressCache = isClustered ? cacheManager.getCache(server.getConfiguration().topologyCacheName()) : null;
      }
      return addressCache;
   }

   @Override
   protected void encode(ChannelHandlerContext ctx, Object msg, ByteBuf buf) throws Exception {
      try {
         if (isTrace) {
            log.tracef("Encode msg %s", msg);
         }

         if (msg instanceof Response) {
            Response r = (Response) msg;
            VersionedEncoder encoder = HotRodVersion.getEncoder(r.version);
            try {
               if (HotRodVersion.forVersion(r.version) != HotRodVersion.UNKNOWN) {
                  encoder.writeHeader(r, buf, getAddressCache(), server);
               } else {
                  // if error before reading version, don't send any topology changes
                  // cos the encoding might vary from one version to the other
                  encoder.writeHeader(r, buf, null, server);
               }

               encoder.writeResponse(r, buf, cacheManager, server);
            } catch (Throwable t) {
               log.errorWritingResponse(r.messageId, t);
               buf.clear(); // reset buffer
               ErrorResponse error = new ErrorResponse(r.version, r.messageId, r.cacheName, r.clientIntel,
                     OperationStatus.ServerError, r.topologyId, t.toString());
               encoder.writeHeader(error, buf, getAddressCache(), server);
               encoder.writeResponse(error, buf, cacheManager, server);
            }
         } else if (msg instanceof Events.Event) {
            Events.Event e = (Events.Event) msg;
            VersionedEncoder encoder = HotRodVersion.getEncoder(e.version);
            encoder.writeEvent(e, buf);
         } else if (msg instanceof ClientCounterEvent) {
            VersionedEncoder encoder = HotRodVersion.getEncoder(((ClientCounterEvent) msg).getVersion());
            encoder.writeCounterEvent((ClientCounterEvent) msg, buf);
         } else if (msg != null) {
            log.errorUnexpectedMessage(msg);
         }
      } catch (Throwable t) {
         log.errorEncodingMessage(msg, t);
         // When the exception is thrown when trying to encode an error response, all we can do
         // is to close the channel to signalize clients that the server has trouble responding.
         ctx.close();
         throw t;
      }
   }
}
